library/ssl_tls.c

```diff
@@ -4685,6 +4689,93 @@ const char *mbedtls_ssl_get_alpn_protocol( const mbedtls_ssl_context *ssl )
 }
 #endif /* MBEDTLS_SSL_ALPN */
 
+#if defined(MBEDTLS_SSL_DTLS_SRTP)
+#if defined(MBEDTLS_DEBUG_C)
+const char *mbedtls_ssl_get_srtp_profile_as_string ( mbedtls_ssl_srtp_profile profile )
+{
+    switch( profile )
+    {
+        case MBEDTLS_TLS_SRTP_AES128_CM_HMAC_SHA1_80:
+            return "MBEDTLS_TLS_SRTP_AES128_CM_HMAC_SHA1_80";
+        case MBEDTLS_TLS_SRTP_AES128_CM_HMAC_SHA1_32:
+            return "MBEDTLS_TLS_SRTP_AES128_CM_HMAC_SHA1_32";
+        case MBEDTLS_TLS_SRTP_NULL_HMAC_SHA1_80:
+            return "MBEDTLS_TLS_SRTP_NULL_HMAC_SHA1_80";
+        case MBEDTLS_TLS_SRTP_NULL_HMAC_SHA1_32:
+            return "MBEDTLS_TLS_SRTP_NULL_HMAC_SHA1_32";
+        default: break;
+    }
+    return( "" );
+}
+#endif /* MBEDTLS_DEBUG_C */
+
+void mbedtls_ssl_conf_srtp_mki_value_supported( mbedtls_ssl_config *conf,
+                                                int support_mki_value )
+{
+    conf->dtls_srtp_mki_support = support_mki_value;
+}
+
+int mbedtls_ssl_dtls_srtp_set_mki_value( mbedtls_ssl_context *ssl,
+                                         unsigned char *mki_value,
+                                         size_t mki_len )
+{
+    if ( mki_len > MBEDTLS_TLS_SRTP_MAX_MKI_LENGTH )
+    {
+        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+    }
+
+    if( ssl->conf->dtls_srtp_mki_support == MBEDTLS_SSL_DTLS_SRTP_MKI_UNSUPPORTED )
+    {
+        return( MBEDTLS_ERR_SSL_FEATURE_UNAVAILABLE );
+    }
+
+    memcpy( ssl->dtls_srtp_info.mki_value, mki_value, mki_len );
+    ssl->dtls_srtp_info.mki_len = mki_len;
+    return( 0 );
+}
+
+int mbedtls_ssl_conf_dtls_srtp_protection_profiles( mbedtls_ssl_config *conf,
+                                                    const mbedtls_ssl_srtp_profile *profiles )
+{
+    const mbedtls_ssl_srtp_profile *p;
+    size_t list_size = 0;
+
+    /* check the profiles list: all entry must be valid,
+     * its size cannot be more than the total number of supported profiles, currently 4 */
+    for( p = profiles; *p != MBEDTLS_TLS_SRTP_UNSET && list_size < 5; p++ )
```

Comment:
What is the reason to restrict the list length to a maximum of `4` here? I see that there's no value in having longer lists, but the code would handle them correctly, and dropping the length check would save a few bytes of code.

Reply:
If you provide a list longer than 4 elements, it means you repeated some element, the order in thus not really defined, which is not really a problem, user shall use correctly the library but the interface should help him too.

Most important is to prevent an infinite loop, or wild memory browsing if the user did not carefully read the instructions and forgot the MBEDTLS_TLS_SRTP_UNSET at the end of its list, some limit must be set anyway, 4 seems to be a good one as you have no real reason to use a longer list.

Reply:
On the other hand, the knowledge that the list length is small would allow to hardcode the high byte of the profile list length in https://github.com/ARMmbed/mbedtls/pull/3235/files#diff-8d3acce410d251c6ead39efd401db23eR805 to `0`.

In any case, this turns into micro-optimization. My preference would be to allow a maximum list length of 255 here and hardcode the high byte in https://github.com/ARMmbed/mbedtls/pull/3235/files#diff-8d3acce410d251c6ead39efd401db23eR805 to `0`, but I leave it to you to decide whether you want to change the code or not. It's not a blocker.

Reply:
If you decide to keep the code as-is, I'd prefer giving the magic values `4` and `5` here a descriptive name, like `MBEDTLS_TLS_SRTP_MAX_PROFILE_LIST_LENGTH` or similar.

Reply:
> If you provide a list longer than 4 elements, it means you repeated some element, the order in thus not really defined, which is not really a problem, user shall use correctly the library but the interface should help him too.
> Most important is to prevent an infinite loop, or wild memory browsing if the user did not carefully read the instructions and forgot the MBEDTLS_TLS_SRTP_UNSET at the end of its list, some limit must be set anyway, 4 seems to be a good one as you have no real reason to use a longer list.

I agree that this is a reasonable way to prevent the user from shooting itself in the foot, but one has to balance such measures with their code size cost. In this instance, the cost is negligible, so I'm ok leaving it. Note, though, that there's a precedent for this, namely the ciphersuite and ALPN configuration API, and that API doesn't do any sanity checking on the list length or content - it's the user's responsibility to get it right.

Reply:
Ok, I kept the check and defined the MAX_PROFILE_LIST_LENGTH.

I also added the micro optimization of writing 0 to the upper byte of the list in the client use_srtp_ext writing code. Note that if you allow the max_length to be more than 127(I kept it at 4), you will need to correctly write the upper byte of the length as you're actually write 2*length in the use_srtp extension